### PR TITLE
Design document

### PR DESCRIPTION
**Pull request by phob0s-pl**

### CODE REVIEW

File: rfd/0000-job-runner.md [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+
+service JobRunner {
+ // starts new job on remote service. The call returns server generated id of the job
+ rpc Start (JobStartRequest) returns (JobStartResponse) {}
+ // stops job scheduled on the server, request must provide valid id, otherwise error will be returned
+ rpc Stop (JobStopRequest) returns (JobStopResponse) {}
+ // info provides verbose information about currently running job
+ rpc Info (JobInfoRequest) returns (Job) {}
+ // list lists all jobs running by the service
+ rpc List (JobListRequest) returns (JobList) {}
```
Comment (russjones) [inline rfd/0000-job-runner.md:34]: This RPC is not required.

File: rfd/0000-job-runner.md [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+ // OutputStream streams stderr and stdout output of running job, firstly it will send historical output, then live stream of current output
+ rpc OutputStream (JobStreamRequest) returns (stream JobStream) {}
+}
+
+message JobInfoRequest {
+ // id of running job
+ string id = 1;
```
Comment (russjones) [inline rfd/0000-job-runner.md:41]: How will you generate this UUID?

File: rfd/0000-job-runner.md [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+  INVALID = 0;
+  // job was received by service
+  PENDING = 1;
+  // job was started and by service and is running
+  RUNNING = 2;
+  // start of a job failed
+  START_FAILED = 3;
+  // process exited with 0 exit code
+  EXITED = 4;
+  // process exited with non 0 exit code
```
Comment (russjones) [inline rfd/0000-job-runner.md:84]: Can you also return the exit code?

File: rfd/0000-job-runner.md [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+### request lifecycle on example of new job request
+1. User invokes command from CLI
+2. client fills gRPC request with token and command line parameters
+3. gRPC adapter from service receives request and validates token using IAM port. Request is translated to domain.Job structure and processed to application logic
+4. Application validates request, check if user is authorized using IAM port and assigns 6 bytes random id, hex encoded
+5. Job is stored in jobs repository port
+6. Job is send to runner port and executed
+7. response is sent to client
+
+### job execution
```
Comment (russjones) [inline rfd/0000-job-runner.md:122]: Can you talk about your process execution model a bit? How will you place your process in a cgroup?

File: rfd/0000-job-runner.md [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+Each job (if has defined cgroups) is configured under ```/sys/fs/cgroup/caes/<job_id>```
+
+This adapter provides actual process execution management.
+- CPU limit is achieved by setting quota and period in ```/sys/fs/cgroup/caes/<job_id>/cpu.max```
+- Memory limit is set in ```/sys/fs/cgroup/caes/<job_id>/memory.max```
+- io limit is set in ```/sys/fs/cgroup/caes/<job_id>/io.max``` for block devices fetched from ```/sys/block/<disk>/dev```
+
+Service is using hardcoded cgroup limits.
+
+Before process is started its stderr/stdout output is set to MBuffer, per process structure which caches output and allows for streaming.
```
Comment (russjones) [inline rfd/0000-job-runner.md:135]: Good that you covered the `io.Writer` interface, can you talk about how `io.Reader` will work as well?

File: rfd/0000-job-runner.md [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+The server architecture consists of:
+1. Domain - which will provide definition of core entities (job, user) and business logic (like: only active user can start a job)
+2. Application - controls flow and defines logic between interface and infrastructure
+3. Interface ports: API port - allows server to use different request sources without affecting application logic
+4. Infrastructure ports: IAM for user authentication and authorization, jobs repository for storing information about job, runner for actual job execution.
+5. Adapters: GRPC adapter is allowing server to be accessible from outside, IAM and jobs repository adapters are in-memory providing minimal functionality. IAM uses OAUTH2 tokens. Local runner adapter is implemented in a way that in only allows processes to be executed on local host.
+
+GRPC adapter is using IAM port for authorization, however authentication is done on application/domain level.
+Edge case scenario testing of application is achieved by writing proper test implementation of adapters, for example: always failing job runner.
```
Comment (russjones) [inline rfd/0000-job-runner.md:109]: Can you explain your authorization scheme. You want to answer the question: who can do what on this system?

File: rfd/0000-job-runner.md (hunk 1 above)
Comment (phob0s-pl) [reply rfd/0000-job-runner.md:34]: I want to keep it for testing CLI by myself

File: rfd/0000-job-runner.md (hunk 2 above)
Comment (phob0s-pl) [reply rfd/0000-job-runner.md:41]: This is defined in section: request lifecycle on example of new job request, point 4:
"assigns 6 bytes random id, hex encoded"

File: rfd/0000-job-runner.md (hunk 3 above)
Comment (phob0s-pl) [reply rfd/0000-job-runner.md:84]: yes

File: rfd/0000-job-runner.md (hunk 5 above)
Comment (phob0s-pl) [reply rfd/0000-job-runner.md:135]: With proposed architecture, connection is going down from grpc adapter -> api port -> application -> job executor -> MBuffer of job, which will do streaming, therefore using io.Writer looks more natural choice for this. Using io.Reader, would require exposing MBuffer to "upper layer" and some reading logic would need to be introduced

File: rfd/0000-job-runner.md (hunk 4 above)
Comment (phob0s-pl) [reply rfd/0000-job-runner.md:149]: Please see: "job execution" section

File: rfd/0000-job-runner.md (hunk 6 above)
Comment (phob0s-pl) [reply rfd/0000-job-runner.md]: Added more verbose definition how authorization will work

File: rfd/0000-job-runner.md [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+  RUNNING = 2;
+  // start of a job failed
+  START_FAILED = 3;
+  // process exited with 0 exit code
+  EXITED = 4;
+  // process exited with non 0 exit code
+  FAILED = 5;
+  // process was stopped by the user
+  STOPPED = 6;
+  // job is present in database, but it was impossible to fetch its current state
```
Comment (jakule) [inline rfd/0000-job-runner.md:88]: Any example of when this can happen?

File: rfd/0000-job-runner.md [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+Following domain models are proposed:
+```go
+// Job contains information about the job
+type Job struct {
+	// ID is unique ID generated by server
+	ID string `json:"id"`
+	// Command is process which will be executed
+    Command          string
+	// Args is list of command arguments
+    Args             []string
```
Comment (jakule) [inline rfd/0000-job-runner.md:122]: nit: looks like formatting is off here. Also, I'm not sure why some fields have `json` tags?
```suggestion
	// Command is process which will be executed
	Command string
	// Args is list of command arguments
	Args []string
```

File: rfd/0000-job-runner.md [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+| token guessing                                                          | authentication | ensure that access token login is long enough to prevent it from guessing                                                                                                                           |
+ | service is running on outdated server configuration                     | devops         | define workflow for execution environment, eg linux version, migration, updates scenario                                                                                                            |
+| weak algorithms are used in certificates                                | authentication | have whitelist of accepted cryptographic algorithms to prevent weak algorithms DES or RC4 to be used                                                                                                |
+ | weak version of SSL/TLS can be used                                     | authentication | ensure supported SSL versions does not have vulnerabilities                                                                                                                                         |
```
Comment (jakule) [inline rfd/0000-job-runner.md:262]: Any recommendations here?

File: rfd/0000-job-runner.md [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+ | user can view output of process started by other user                   | authorization  | define security policy if such a behavior is correct                                                                                                                                                |
+ | server runs out of available file descriptors                           | transport      | provide timeout for idle connections                                                                                                                                                                |
+| generated job id collides with running one                              | application    | use longer job ID, verify against running ones and regenerate if needed                                                                                                                             |
+| authorization credentials cannot be send in insecure channel            | authentication | enforce using TLS certificates by client and server                                                                                                                                                 |
```
Comment (jakule) [inline rfd/0000-job-runner.md:257]: I think this is already covered by your solution.

File: rfd/0000-job-runner.md [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+
+```
+
+![architecture](architecture.png)
+
+### request lifecycle on example of new job request
+1. User invokes command from CLI
+2. client fills gRPC request with token and command line parameters
+3. gRPC adapter from service receives request and validates token using IAM port. Request is translated to domain.Job structure and processed to application logic
+4. Application validates request, check if user is authorized using IAM port and assigns 6 bytes random id, hex encoded
```
Comment (jakule) [inline rfd/0000-job-runner.md:144]: Out of curiosity, why 6 bytes?

File: rfd/0000-job-runner.md [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+```
+
+### Server
+The server is following hexagonal architecture (known also as ports and adapters). This approach follows good coding practice, provides maintainability and allows to test each component separately.
+The server architecture consists of:
+1. Domain - which will provide definition of core entities (job, user) and business logic (like: only active user can start a job)
+2. Application - controls flow and defines logic between interface and infrastructure
+3. Interface ports: API port - allows server to use different request sources without affecting application logic
+4. Infrastructure ports: IAM for user authentication and authorization, jobs repository for storing information about job, runner for actual job execution.
+5. Adapters: GRPC adapter is allowing server to be accessible from outside, IAM and jobs repository adapters are in-memory providing minimal functionality. IAM uses OAUTH2 tokens. Local runner adapter is implemented in a way that in only allows processes to be executed on local host.
```
Comment (jakule) [inline rfd/0000-job-runner.md:107]: Why do we need tokens?

File: rfd/0000-job-runner.md (hunk 7 above)
Comment (phob0s-pl) [reply rfd/0000-job-runner.md:88]: This process execution is separated from execution with runner port, meaning that process execution can be done remotely in future. That means that process execution port needs to return error when calling ```runner.GetJobStatus(id string) (Status, err)``` The UNKNOWN state is when: application fetches successfully domain.Job from database port, but there was timeout/communication issue when communicating with runner port.

File: rfd/0000-job-runner.md (hunk 11 above)
Comment (phob0s-pl) [reply rfd/0000-job-runner.md:141]: Balance between user experience ```caes info <super long id>``` is something end user may not like, and having acceptable collision rate for this PoC. I have covered collision topic in security concern.
Stolen length from docker id :)

File: rfd/0000-job-runner.md (hunk 10 above)
Comment (phob0s-pl) [reply rfd/0000-job-runner.md]: true

File: rfd/0000-job-runner.md (hunk 9 above)
Comment (phob0s-pl) [reply rfd/0000-job-runner.md]: added recommendations

File: rfd/0000-job-runner.md (hunk 8 above)
Comment (phob0s-pl) [reply rfd/0000-job-runner.md]: fixed formatting

File: rfd/0000-job-runner.md [hunk 13]
```diff
@@ -0,0 +1,305 @@
...
+message JobInfoRequest {
+ // id of running job
+ string id = 1;
+}
+
+message JobStreamRequest {
+ string id = 1;
+}
+
+message JobStream {
```
[the remark this answers is not in the document]
Comment (phob0s-pl) [reply rfd/0000-job-runner.md]: applied those changes

File: rfd/0000-job-runner.md [hunk 14]
```diff
@@ -0,0 +1,305 @@
...
+  // job was received by service
+  PENDING = 1;
+  // job was started and by service and is running
+  RUNNING = 2;
+  // start of a job failed
+  START_FAILED = 3;
+  // process exited with 0 exit code
+  EXITED = 4;
+  // process exited with non 0 exit code
+  FAILED = 5;
```
[the remark this answers is not in the document]
Comment (phob0s-pl) [reply rfd/0000-job-runner.md]: You are right, once I have introduced exit code, I didn't notice that having failed does not make sense

File: rfd/0000-job-runner.md (hunk 4 above)
Comment (russjones) [reply rfd/0000-job-runner.md:146]: Maybe I am missing something, but I don't see where you outline how a process is placed in a cgroup?

File: rfd/0000-job-runner.md (hunk 5 above)
Comment (russjones) [reply rfd/0000-job-runner.md:160]: Sorry, I don't fully understand. It's okay if you don't want to use a `io.Reader` abstraction, but what do you plan to use instead?

Comment (phob0s-pl) [reply rfd/0000-job-runner.md:160]: What I want to have is MBuffer, which:
- as input will provide io.Writer interface. This will be assigned to exec.Cmd  stderr and stdout of assigned process
- as output of MBuffer I see many dynamic  io.WriteClosers of streaming connection right now, but during implementation I will see if  io.Reader is better  ( never done gRPC streaming before )

File: rfd/0000-job-runner.md [hunk 15]
```diff
@@ -0,0 +1,301 @@
...
+- CPU limit is achieved by setting quota and period in ```/sys/fs/cgroup/caes/<job_id>/cpu.max```
+- Memory limit is set in ```/sys/fs/cgroup/caes/<job_id>/memory.max```
+- io limit is set in ```/sys/fs/cgroup/caes/<job_id>/io.max``` for block devices fetched from ```/sys/block/<disk>/dev```
+
+Service is using hardcoded cgroup limits.
+
+Before process is started its stderr/stdout output is set to MBuffer, per process structure which caches output and allows for streaming.
+MBuffer has two components internal buffer and stream logic. Each job writes output to buffer using ```io.Writer``` interface.
+Client application which requests stream first receives historical job output, then is moved to stream logic, where it receives live stream.
+There is consideration to make client writers as ```io.WriteCloser```, meaning that when job exits/is stopped, WriteClosers then are closed, which would lead to closing connection instead of keeping it open.
```
[the remark this answers is not in the document]
Comment (phob0s-pl) [reply rfd/0000-job-runner.md:163]: By calling Close() if I go io.WriteCloser way, or return EOF it io.Reader will turn out to be better

File: rfd/0000-job-runner.md (hunk 4 above)
Comment (phob0s-pl) [reply rfd/0000-job-runner.md:170]: There is "Process PID is added to /sys/fs/cgroup/caes/<job_id>/cgroup.procs" :)

File: rfd/0000-job-runner.md (hunk 12 above)
Comment (phob0s-pl) [reply rfd/0000-job-runner.md]: Discarded token approach

File: rfd/0000-job-runner.md [hunk 16]
```diff
@@ -0,0 +1,301 @@
...
+2. Application - controls flow and defines logic between interface and infrastructure
+3. Interface ports: API port - allows server to use different request sources without affecting application logic
+4. Infrastructure ports: IAM for user authentication and authorization, jobs repository for storing information about job, runner for actual job execution.
+5. Adapters: GRPC adapter is allowing server to be accessible from outside, IAM and jobs repository adapters are in-memory providing minimal functionality. IAM uses OAUTH2 tokens. Local runner adapter is implemented in a way that in only allows processes to be executed on local host.
+
+GRPC adapter is using IAM port for authorization, however authentication is done on application/domain level.
+The authentication will be done in following way: IAM port provides interface to get user by token```GetUserByToken(token string) (*domain.User, error)``` , when user structure is known basic authorization rule is applied: if user is active, he can start job. This will also demonstrate how business logic is handled by application.
```
[the remark this answers is not in the document]
Comment (phob0s-pl) [reply rfd/0000-job-runner.md]: Discarded token approach

File: rfd/0000-job-runner.md [hunk 17]
```diff
@@ -0,0 +1,301 @@
...
+	Active bool
+	Token  string
+}
+```
+
+![architecture](architecture.png)
+
+### request lifecycle on example of new job request
+1. User invokes command from CLI
+2. client fills gRPC request with token and command line parameters
```
[the remark this answers is not in the document]
Comment (phob0s-pl) [reply rfd/0000-job-runner.md]: I have created scheme which does not need tokens anymore

File: rfd/0000-job-runner.md [hunk 18]
```diff
@@ -0,0 +1,301 @@
...
+3. gRPC adapter from service receives request and validates token using IAM port. Request is translated to domain.Job structure and processed to application logic
```
[the remark this answers is not in the document]
Comment (phob0s-pl) [reply rfd/0000-job-runner.md]: I have created scheme which does not need tokens anymore. But will keep IAM port to have modularity ( eg. to have possiility to make FusionAuth adapter in the future )

File: rfd/0000-job-runner.md [hunk 19]
```diff
@@ -0,0 +1,301 @@
...
+5. Job is stored in jobs repository port
+6. Job is send to runner port and executed
+7. response is sent to client
+
+### job execution
+For namespace creation proper flags are set in ```SysProcAttr``` of ```exec.Cmd``` structure.
+Mount namespace will provide job unique rootfs.
+Job runner is using cgroups v2, during startup root group ```caes``` is created in ```/sys/fs/cgroup/caes```.
+Each job (if has defined cgroups) is configured under child cgroup ```/sys/fs/cgroup/caes/<job_id>```
+Process PID is added to ```/sys/fs/cgroup/caes/<job_id>/cgroup.procs```
```
[the remark this answers is not in the document]
Comment (phob0s-pl) [reply rfd/0000-job-runner.md]: Added when exec.Cmd is started

File: rfd/0000-job-runner.md [hunk 20]
```diff
@@ -0,0 +1,301 @@
...
+    input_writer-->buffer;
+    input_writer-->stream;
+    buffer-->client_a_writer;
+    buffer-->client_b_writer;
+    stream-->client_a_writer;
+    stream-->client_b_writer;
+```
+
+### mTLS
+caes is using mTLS for authentication. To achieve this following certificates and keys must be present: CA certificate, client private key, client certificate, server private key and server certificate.
```
[the remark this answers is not in the document]
Comment (phob0s-pl) [reply rfd/0000-job-runner.md:202]: added :)

File: rfd/0000-job-runner.md [hunk 21]
```diff
@@ -0,0 +1,301 @@
...
+ // id of running job
+ string id = 1;
+}
+
+message JobStreamRequest {
+ string id = 1;
+}
+
+message JobStreamResponse {
+ bytes data = 1;
```
[the remark this answers is not in the document]
Comment (phob0s-pl) [reply rfd/0000-job-runner.md:50]: Its both. It will work like cmd.CombinedOutput()

File: rfd/0000-job-runner.md [hunk 22]
```diff
@@ -0,0 +1,301 @@
...
+This adapter provides actual process execution management.
+- CPU limit is achieved by setting quota and period in ```/sys/fs/cgroup/caes/<job_id>/cpu.max```
+- Memory limit is set in ```/sys/fs/cgroup/caes/<job_id>/memory.max```
+- io limit is set in ```/sys/fs/cgroup/caes/<job_id>/io.max``` for block devices fetched from ```/sys/block/<disk>/dev```
+
+Service is using hardcoded cgroup limits.
+
+Before process is started its stderr/stdout output is set to MBuffer, per process structure which caches output and allows for streaming.
+MBuffer has two components internal buffer and stream logic. Each job writes output to buffer using ```io.Writer``` interface.
+Client application which requests stream first receives historical job output, then is moved to stream logic, where it receives live stream.
```
[the remark this answers is not in the document]
Comment (phob0s-pl) [reply rfd/0000-job-runner.md:186]: It's output which a) is already stored in buffer b) one needed to be send to connection between live

File: rfd/0000-job-runner.md [hunk 23]
```diff
@@ -0,0 +1,350 @@
...
+	cfg := &tls.Config{
+	    //tradeoff: root CA is used for client cert generation
+		ClientCAs:        CaCertPool,
+		RootCAs:          CaCertPool,
+		ClientAuth:       tls.RequireAndVerifyClientCert,
+		Certificates:     []tls.Certificate{serverCertificate},
+		MinVersion:       tls.VersionTLS13,
+		MaxVersion:       tls.VersionTLS13,
+		// function called on every connection which will verify certificate details
+		VerifyConnection: VerifyConnection,
```
[the remark this answers is not in the document]
Comment (phob0s-pl) [reply rfd/0000-job-runner.md:217]: It depends how big constrains are put to certificate

File: rfd/0000-job-runner.md [hunk 24]
```diff
@@ -0,0 +1,350 @@
...
+        return true
+    }
+    return false      
+}
+```
+
+### job execution
+Job runner is using cgroups v2, during startup root group ```caes``` is created in ```/sys/fs/cgroup/caes``` if it does not exist.
+For namespace creation proper flags are set in ```SysProcAttr``` of ```exec.Cmd``` structure.
+Process is executed with ```exec.Cmd.Start()``` function, separate goroutine is created to wait until process will finish.
```
[the remark this answers is not in the document]
Comment (phob0s-pl) [reply rfd/0000-job-runner.md:173]: Will make research about this,thanks!

File: rfd/0000-job-runner.md [hunk 25]
```diff
@@ -0,0 +1,350 @@
...
+		Certificates:     []tls.Certificate{serverCertificate},
+		MinVersion:       tls.VersionTLS13,
+		MaxVersion:       tls.VersionTLS13,
+		// function called on every connection which will verify certificate details
+		VerifyConnection: VerifyConnection,
+	}
+```
+
+
+### CLI
```
[the remark this answers is not in the document]
Comment (phob0s-pl) [reply rfd/0000-job-runner.md:222]: There is ```--addr``` flag, adding to document

File: rfd/0000-job-runner.md [hunk 26]
```diff
@@ -0,0 +1,350 @@
...
+This adapter provides actual process execution management.
+- CPU limit is achieved by setting quota and period in ```/sys/fs/cgroup/caes/<job_id>/cpu.max```
+- Memory limit is set in ```/sys/fs/cgroup/caes/<job_id>/memory.max```
+- io limit is set in ```/sys/fs/cgroup/caes/<job_id>/io.max``` for block devices fetched from ```/sys/block/<disk>/dev```
+
+Service is using hardcoded cgroup limits.
+
+Before process is started its stderr/stdout output is set to MBuffer, per process structure which caches output and allows for streaming.
+Each job writes output to buffer using ```io.Writer``` interface.
+MBuffer provides ```io.reader``` through ```mbuffer.CreateReader()```  created per stream. When there will be no more data available (process exited) ```EOF``` is returned.
```
[the remark this answers is not in the document]
Comment (phob0s-pl) [reply rfd/0000-job-runner.md]: Added io.ReadCloser approach. When connection will be closed, server will propagate the news to MBuffer by calling Close on io.ReadCloser. But I will see pitfails during implementation :)

File: rfd/0000-job-runner.md (hunk 25 above)
[the remark this answers is not in the document]
Comment (phob0s-pl) [reply rfd/0000-job-runner.md:226]: yes, added to document

File: rfd/0000-job-runner.md (hunk 23 above)
[the remark this answers is not in the document]
Comment (phob0s-pl) [reply rfd/0000-job-runner.md:221]: Previously I was asked to propose acceptable algorithms.
VerifyConnection is a place where such a parameters of certificate can be checked by callback function, which can look like this:
```go
func verifyConnection(cs tls.ConnectionState) error {
	if len(cs.PeerCertificates) == 0 {
		return errMissingCertificate
	}

	switch cs.PeerCertificates[0].SignatureAlgorithm {
	case x509.SHA512WithRSA, x509.SHA256WithRSA, x509.ECDSAWithSHA512:
	default:
		return errInvalidCertificate
	}

	return nil
}
```

[the remark this answers is not in the document]
Comment (phob0s-pl) [reply rfd/0000-job-runner.md:221]: Other thing which came to my mind is to check ( thing mentioned in security consideration section ):
It can be useful for validating if in system where certificates are rotated often, they have short lifetime:

```go
func verifyConnection(cs tls.ConnectionState) error {
	if len(cs.PeerCertificates) == 0 {
		return errMissingCertificate
	}

       // pseudocode
	if cs.PeerCertificates[0].NotAfter.Sub(cs.PeerCertificates[0].NotBefore) != time.Hour*24 {
		return errInvalidCertificateDucration
	}

	return nil
}
```

[the remark this answers is not in the document]
Comment (jakule) [reply rfd/0000-job-runner.md:221]: Why are you specifically looking for a certificate valid exactly 24 hours? Do you want to check the certificate validity instead of signature algorithm or you want to check both of them? What type of potential attack those checks can prevent?

**Review by russjones — APPROVED** [review]

**Review by jakule — APPROVED** [review]
